packages/runner-ct/src/app/SpecList/SpecList.tsx

```diff
@@ -0,0 +1,393 @@
+        indexes={indexes}
+        item={item}
+        onClick={props.onFileClick}
+      />
+    )
+
+    // if no search entered we just show all the specs.
+    if (props.search === undefined) {
+      return render([])
+    }
+
+    const match = checkMatch(item, props.matches)
+
+    if (!match) {
+      return <span />
+    }
+
+    return render(match.indexes)
+  }
+
+  return (
+    <>
+      <ul className={styles && styles.ul}>
+        {
+          props.files.map((item) => {
+            return (
+              <React.Fragment key={item.relative}>
+                <a
+                  data-item={item.relative}
+                  style={{
+                    marginLeft: `-${20 * props.depth}px`,
+                    width: `calc(100% + (20px * ${props.depth}))`,
+                  }}
+                  className={cs(styles.a, {
+                    [styles.isSelected]: item.relative === props.selectedFile,
+                  })}
+                  tabIndex={0}
+                >
+                  <li
+                    style={{ marginLeft: `${20 * props.depth}px` }}
+                    className={styles.li}>
+                    {item.type === 'folder' ? renderFolder(item) : renderFile(item)}
+                  </li>
+                </a>
+                {fileTree(item)}
+              </React.Fragment>
+            )
+          })
+        }
+      </ul>
+    </>
+  )
+}
+
+export const SpecList: React.FC<SpecListProps> = (props) => {
+  const files = React.useMemo(() => makeFileHierarchy(props.specs.map((spec) => spec.relative)), [props.specs])
+
+  /**
+   * Whether a folder is open or not is a **UI** concern.
+   * From a file system point of view, there is no such concept as "open" or "closed",
+   * only from a user's point of view.
+   * For this reason we save the open state as part of the UI component. The easiest
+   * way to do this is a key/value pair, mapping the relative path of a directory to a boolean
+   *
+   * {
+   *   'foo': true,
+   *   'foo/bar': true
+   *   'foo/bar/qux': false
+   * }
+   *
+   * Every directory is set to open by default. When you add a new directory
+   * or file via your file system (eg mkdir foo/bar && touch foo/bar/hello.js) it will be added
+   * without losing the current state of open/closed directories.
+   */
+  const [openFolders, setOpenFolders] = React.useState<Record<string, boolean>>({})
+  const [search, setSearch] = React.useState('')
+
+  const fuzzyTransform = (spec: Cypress.Cypress['spec'], indexes: number[]) => {
+    const split = spec.relative.split('/')
+    const name = split[split.length - 1]
+
+    return {
+      relative: spec.relative,
+      name,
+      indexes,
+    }
+  }
+
+  const matches = useMemo(() => {
+    if (!search) {
+      return props.specs.map((spec) => fuzzyTransform(spec, []))
+    }
+
+    return fuzzysort.go(search, props.specs, { key: 'relative' }).map((result) => fuzzyTransform(result.obj, result.indexes))
```

Comment:
^ you should not think about performance problems until they are really a problem. Using useMemo is much more declarative. Moreover, if you will go using `useEffect` it will end up with 2 commit phases and 2 dom updates which is not needed here.

Reply:
I would argue that 2 commit phases is vastly preferred. `SearchSpec` is a controlled component, so there will be no character rendered in the text box until the `fuzzysort` finishes. My experience with this library is that you start noticing the performance issue very quickly.

Reply:
I'm interested in writing this hook either way, so I will give it a try and see how it looks.

Reply:
I updated this to have a `useFuzzySort` hook. I will need a review on it by @agg23. I am not entirely convinced the in the premature optimization but happy to include it if we are expecting perf issues. Please review again @agg23.

PS: there is an infinite loop if I include all the dependences in the hook (I left a TODO). Do you know why this is happening?